### PR TITLE
Enable returning a custom status code from the Adaptive Concurrency Filter

### PR DESCRIPTION
**Pull request by RiverPhillips**
Commit Message: Enable returning a custom status code from the Adaptive Concurrency Filter
Additional Description: Like global rate limit will not allow blocking with a successful status code, if an invalid status code is encountered it will fall back to 503.
Risk Level: Low
Testing: Unit tests added
Docs Changes: proto docs added
Release Notes: n/a
Platform Specific Features: n/a
Fixes: #23696

### CODE REVIEW

File: api/envoy/extensions/filters/http/adaptive_concurrency/v3/adaptive_concurrency.proto
Comment (adisuissa) [inline api/envoy/extensions/filters/http/adaptive_concurrency/v3/adaptive_concurrency.proto:111]: nit:
```suggestion
  // the concurrency limit has been exceeded.
```

File: api/envoy/extensions/filters/http/adaptive_concurrency/v3/adaptive_concurrency.proto
Comment (adisuissa) [inline api/envoy/extensions/filters/http/adaptive_concurrency/v3/adaptive_concurrency.proto:112]: nit:
```suggestion
  // Defaults to 503 (Service Unavailable).
```

File: source/extensions/filters/http/adaptive_concurrency/adaptive_concurrency_filter.h
Comment (adisuissa) [inline source/extensions/filters/http/adaptive_concurrency/adaptive_concurrency_filter.h:40]: nit: Please move to an unnamed namespace inner function in the cc file.

File: test/extensions/filters/http/adaptive_concurrency/adaptive_concurrency_filter_test.cc
Comment (adisuissa) [inline test/extensions/filters/http/adaptive_concurrency/adaptive_concurrency_filter_test.cc:409]: Please add a short comment before the tests describing them.

**Review by adisuissa — CHANGES_REQUESTED** [review]
Thanks for adding this feature!
Overall looks good. Left a few minor comments.
/wait

File: source/extensions/filters/http/adaptive_concurrency/adaptive_concurrency_filter.h
Comment (RiverPhillips) [reply source/extensions/filters/http/adaptive_concurrency/adaptive_concurrency_filter.h:40]: I'm not super familiar with modern C++, in fact this is the first bit of C++ I've written since undergrad. Can you clarify what you mean by an "unnamed namespace inner function" or link to an example?

**Comment by RiverPhillips** [comment]
Thanks for the feedback, I'm on holiday and will try to address this when I return early next week

File: source/extensions/filters/http/adaptive_concurrency/adaptive_concurrency_filter.h
Comment (adisuissa) [reply source/extensions/filters/http/adaptive_concurrency/adaptive_concurrency_filter.h:40]: Sure. Currently `toErrorCode` is a function that is only used in the cc file, and is also defined as private (shouldn't be used by other classes).
In this case we can move the function into the cc file inside an unnamed namespace, e.g.,
```
namespace {
  Http::Code toErrorCode(...) {
    ...
  }
}
```

Example can be seen [here](https://github.com/envoyproxy/envoy/blob/a2b595da4fd48c7c6fa399637f5001d0c0f0f6a5/source/extensions/filters/http/oauth2/config.cc#L25-L38) where `secretsProvider()` is defined only internally in this file.

**Comment by RiverPhillips** [comment]
Should be resolved now

File: api/envoy/extensions/filters/http/adaptive_concurrency/v3/adaptive_concurrency.proto
Comment (adisuissa) [inline api/envoy/extensions/filters/http/adaptive_concurrency/v3/adaptive_concurrency.proto:116]: Sorry, one more question that I missed before, why is this in `AdaptiveConcurrency` and not in `GradientControllerConfig`?

Comment (RiverPhillips) [reply api/envoy/extensions/filters/http/adaptive_concurrency/v3/adaptive_concurrency.proto:116]: I tried to keep adaptive concurrency consistent with `RateLimit`. The 503 was returned by `AdaptiveConcurrency` so I thought it made sense to configure the status code at this level. I got the impression that the `GradientControllerConfig` was related more to calculating the value than handling the response.

Comment (RiverPhillips) [reply api/envoy/extensions/filters/http/adaptive_concurrency/v3/adaptive_concurrency.proto:116]: I'm happy either way, made sense for me to put it here as that's where the existing 503 was returned but a second opinion can't hurt

Comment (RiverPhillips) [reply api/envoy/extensions/filters/http/adaptive_concurrency/v3/adaptive_concurrency.proto:116]: I looked at refactoring, at the moment the `GradientController` doesn't have any HTTP level concepts in it. It seems a bit strange to me to  introduce it for this use case

Comment (adisuissa) [reply api/envoy/extensions/filters/http/adaptive_concurrency/v3/adaptive_concurrency.proto:116]: The way I read it, it seems that GradientController is currently a type of HTTP adaptive controller (could be others in the future), and that it imposes the limits.
As it also defines the concurrency limits, it may also define the return val.
That said, we can probably have this field as a filter level setting, like you are suggesting, because it points out to a configuration of the HTTP filter side of things. I guess we can have an override for this in the future, if additional controllers are added.

**Review by adisuissa — COMMENTED** [review]
/lgtm api
/assign-from @envoyproxy/senior-maintainers 

Please add a release note for the new feature.

**Review by lizan — APPROVED** [review]
LGTM, defer merging to @adisuissa when release note is ready.

**Comment by RiverPhillips** [comment]
Hi @adisuissa is there anything I can do to help move this forward?

**Comment by alyssawilk** [comment]
@adisuissa ping!

**Review by adisuissa — APPROVED** [review]
Apologies for the late reply.
/lgtm api

**Comment by adisuissa** [comment]
Sorry, just saw that @lizan already approved. Merging.

**Pull request merged**
